### ISSUE TITLE
Compile on Windows with Visual Studio

### ISSUE THREAD

**Issue by AndreV2890** [terzakig/gptam#3]
Hi,

i'd like to try your version of PTAM but i need to compile the source code for windows with visual studio 2015.
Can I use the same CMakeLists.txt file? I think that I've to change more or less two parts:
DEFINING NECESSARY MACROS (but i do not know how) and where you define the path of the opencv inlude and libs.

**Comment by terzakig** [comment]
Hi Andrea,


Thanks for your interest in this piece of work.


I can guarantee that the code that does image processing, reconstruction, and bundle adjustment (i.e., the actual PTAM) can compile on windows because it relies solely on OpenCV.


However, in terms of the GUI, I mostly kept the original graphical interface (with modifications) which is OpenGL and it may need slight modifications because I recall Klein made use of a few functions that are available only in Linux; but you can replace them with glut equivalents. Also, as far as I recall, there is also a header (see "Persistence" folder ) for parsing strings that exists only in Linux, but you can actually write your own code for it or simply not use it at all. In general, you may encounter small problems with code pertinent to the graphical interface and the real-time parameter retrieval from files, but you can fix them as you go. The rest is based on OpenCV so, as long as it is correctly installed, it should compile.



So, I would say you can try to compile with the CMakeLists.txt provided and take the errors one at a time. Although I never tried it, I think Visual studio should create a solution without problems.


One suggestion: Since there are essentially two projects, the calibrator and PTAM, I would suggest to try to isolate the calibrator files and work on them. If it compiles, then you should be able to compile PTAM the same way.


Hope this is helpful.


Best


George


________________________________

**Comment by AndreV2890** [comment]
Hi George thanks for your immediate response

> Although I never tried it, I think Visual studio should create a solution without problems.

Using the CMake GUI it is possible to create an MV2015 project. Obviously, the section where you include the OpenCV must be modified, but it is trivial.

> However, in terms of the GUI, I mostly kept the original graphical interface (with modifications) which is OpenGL and it may need slight modifications because I recall Klein made use of a few functions that are available only in Linux; but you can replace them with glut equivalents. 

I don't know if I'm interested in the PTAM GUI.
In my application, I receive the streaming video from a Raspberry and I'd like to apply the PTAM tracking on that.  Is the GUI mandatory or not? Can I show the tracked points and the map for instance in simple OpenCV windows?


> Also, as far as I recall, there is also a header (see "Persistence" folder ) for parsing strings that exists only in Linux, but you can actually write your own code for it or simply not use it at all.

Probably I will skip these files.

**Comment by terzakig** [comment]
Hi,

It seems I have missed one question in your last message. Apologies for that and hope is not extremely too late for my response. here it is, in case you are still interested:

No, the GUI is definitely not necessary to run the application. Visualizations are up to you. however, there are several controls used to interface with PTAM. These are for capturing calibration images, reseting the calibrator and PTAM, initializaing SLAM, quitting the application and running the calibration iteration. They are not many, but you will have to find a way to rigger these operations through your own interface (which does not have to be graphical necessarily). 

Now, in terms of visualizations, you can use the OpenCV routines to draw the tracked points and the camera location along with the reconstruction of the scene.That being said, you will have to remove the AR visualizations with OpenGL. If you are only interested in the SLAM side of things, then this is actually a good idea.

The above are actually simplifications and they should lighten the code, but you would have to study the code a bit.

So, in short, you can actually isolate the tracker, the mapmaker, the patchfinder (along with several other  pertinent classes) and use them in your own visual SLAM program with an OpenCV based interface (see ORB SLAM). The required modifications should not be extensive, but you will have to do it carefully.
